spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/PushDown.scala

```diff
@@ -0,0 +1,204 @@
+  def predicateColumnName(filter: Filter) : String = {
+    filter match {
+      case eq: sources.EqualTo            => eq.attribute
+      case lt: sources.LessThan           => lt.attribute
+      case le: sources.LessThanOrEqual    => le.attribute
+      case gt: sources.GreaterThan        => gt.attribute
+      case ge: sources.GreaterThanOrEqual => ge.attribute
+      case in: sources.In                 => in.attribute
+      case _ =>
+        throw new UnsupportedOperationException(
+          "It's not a valid filter to be pushed down, only >, <, >=, <= and In are allowed: " + filter)
+    }
+  }
+}
+
+/**
+ * Calculate the pushdown predicates for a given table and predicates.
+ *
+ * Partition pruning predicates are also detected an applied.
+ *  1. Only push down no-partition key column predicates with =, >, <, >=, <= predicate
+ *  2. Only push down primary key column predicates with = or IN predicate.
+ *  3. If there are regular columns in the pushdown predicates, they should have
+ *     at least one EQ expression on an indexed column and no IN predicates.
+ *  4. All partition column predicates must be included in the predicates to be pushed down,
+ *     only the last part of the partition key can be an IN predicate. For each partition column,
+ *     only one predicate is allowed.
+ *  5. For cluster column predicates, only last predicate can be non-EQ predicate
+ *     including IN predicate, and preceding column predicates must be EQ predicates.
+ *     If there is only one cluster column predicate, the predicates could be any non-IN predicate.
+ *  6. There is no pushdown predicates if there is any OR condition or NOT IN condition.
+ *  7. We're not allowed to push down multiple predicates for the same column if any of them
+ *     is equality or IN predicate.
+ *
+ */
+abstract class PushDown[T](predicates: Seq[T], tableDef: TableDef) {
+
+  /** Check if the predicate is a Eqaul To predicate */
+  def isEqualTo(predicate: T): Boolean
+
+  /** Check if the predicate is In predicate */
+  def isIn(predicate: T) : Boolean
+
+  /** Check if the predicate is a range comparison predicate */
+  def isRangeComparison(predicate: T) : Boolean
+
+  /** Check if the column is a single column predicate */
+  def isSingleColumn(predicate: T) : Boolean = isEqualTo(predicate) ||  isIn(predicate) || isRangeComparison(predicate)
+
+  /** Returns the only column name referenced in the predicate */
+  def predicateColumnName(predicate: T) : String
+
+  private val partitionKeyColumns = tableDef.partitionKey.map(_.columnName.toLowerCase)
+  private val clusteringColumns = tableDef.clusteringColumns.map(_.columnName.toLowerCase)
+  private val indexedColumns = tableDef.allColumns.filter(_.isIndexedColumn).map(_.columnName.toLowerCase)
+  private val regularColumns = tableDef.regularColumns.map(_.columnName.toLowerCase)
+  private val allColumns = partitionKeyColumns ++ clusteringColumns ++ regularColumns
```

Comment:
But allColumns is already in the tableDef. Do you mean adding more "special case" methods to tableDef? So we'd have "partitionKeyColumns", "partitionKeyColumnNames", "partitionKeyColumnNamesInLowerCase" fields? :D